docs/design/proactive-diagnosis.md

```diff
@@ -0,0 +1,113 @@
+training operations (it can only cover some common cases).
+
+Therefore, a comprehensive, scalable framework is needed to gradually improve 
+the diagnosis and handling of corresponding exception scenarios.
+
+## Target
+
+- Be capable of proactive probing.
+- Requires flexible and scalable implementation.
+- Be able to reuse the current fault tolerance mechanisms.
+
+## Design
+
+The proactive diagnostic framework is mainly composed of the following roles:
+- [Diagnosis Manger](#Diagnosis-Manger)
+- [Diagnosis Agent](#Diagnosis-Agent)
+
+These components are distributed across the Master and Worker Nodes, 
+working together to achieve the main functionality of proactive diagnostics.
+
+<img src="../figures/diagnosis_arc.png" alt="Proactive Diagnostic Framework">
+
+Below is a more detailed introduction to the relevant concepts.
+
+### Diagnosis Manger
+The diagnosis-manager is implemented and runs in the Master node, with the main 
+responsibilities of driving pre-execution checks and runtime diagnostic logic(
+across all the workers).
+
+Runtime diagnostics primarily rely on an asynchronous scheduled thread and a 
+list of potential issues along with corresponding solutions(currently hardcoded). 
+During the training progress, this system continuously performs diagnostics and 
+determines the necessary actions to be executed. The specific diagnostic process 
+is shown in the following diagram, which is implemented through a simple inference chain. 
+The final actions can be either an event output or a specific fault-tolerance operation.
+
+<img src="../figures/diagnosis_manager_arc.png" alt="Diagnosis Manager">
+
+Then it will interact with other core roles of the master, such as the 
+job-manager, to assist in the lifecycle management of training.
+
+### Diagnosis Agent
+The diagnosis-agent is implemented and runs on the worker node, with the main 
+responsibilities of driving data collection on the training worker side, 
+executing some basic diagnostic logic, and reporting necessary information to 
+Master.
+
+<img src="../figures/diagnosis_agent_arc.png" alt="Diagnosis Agent">
+
+### Other Concepts
+#### Metric Collection
+
+- Common Metric
+
+  
+    Common metrics mainly refer to metric information collected directly using 
+    common tools such as psutil or nvml.
+
+- XPU_TIMER Metric
+    
+
+    XPU timer metrics refer to the metrics output by the XPU-Timer. 
+    The collection of this metric is indirectly achieved by using an HTTP 
+    client to access the data exported by the XPU-Timer to Prometheus.
+
+
+## Implementation
+Next, we will primarily introduce some of the existing implementations based on 
+this diagnostic framework. 
+
+This part of the work will have a long cycle and work in progress for now...
+
+### Training Hang Detection(Basic)
+
+```
+prerequisite: XPU-Timer is deployed and in used.
+```
+
+Currently, the basic capability is to determine whether a Hang situation exists. 
```

Comment:
Can we distinguish the hang  checkpoint or dataloader  hang with NCCL OP hang?

Reply:
The XPU-Timer is still gradually being optimized in determining whether it is waiting for a barrier or experiencing an NCCL hang.